lib/core/utils/selector-cache.js

```diff
@@ -0,0 +1,210 @@
+import { matchesExpression } from './matches';
+import tokenList from './token-list';
+
+// since attribute names can't contain whitespace, this will be
+// a reserved list for ids so we can perform virtual id lookups
+const idsKey = ' [idsMap]';
+
+/**
+ * Get nodes from the selector cache that match the selector.
+ * @param {VirtualTree[]} domTree flattened tree collection to search
+ * @param {Object} expressions
+ * @param {Function} filter function (optional)
+ * @return {Mixed} Array of nodes that match the selector or undefined if the selector map is not setup
+ */
+export function getNodesMatchingExpression(domTree, expressions, filter) {
+  // check to see if the domTree is the root and has the selector
+  // map. if not we just return and let our QSA code do the finding
+  const selectorMap = domTree[0]._selectorMap;
+  if (!selectorMap) {
+    return;
+  }
+
+  const shadowId = domTree[0].shadowId;
+
+  // if the selector uses a global selector with a combinator
+  // (e.g. A *, A > *) it's actually faster to use our QSA code than
+  // getting all nodes and using matchesExpression
+  for (let i = 0; i < expressions.length; i++) {
+    if (
+      expressions[i].length > 1 &&
+      expressions[i].some(expression => isGlobalSelector(expression))
+    ) {
+      return;
+    }
+  }
+
+  // it turned out to be more performant to use a Set to generate a
+  // unique list of nodes rather than an array and array.includes
+  // (~3 seconds total on a benchmark site)
+  const nodeSet = new Set();
+
+  expressions.forEach(expression =>
+    findMatchingNodes(expression, selectorMap, shadowId, nodeSet)
+  );
+
+  // Sets in ie11 do not work with Array.from without a polyfill
+  //(missing `.entries`), but do have forEach
+  let matchedNodes = [];
+  nodeSet.forEach(node => matchedNodes.push(node));
+
+  if (filter) {
+    matchedNodes = matchedNodes.filter(filter);
+  }
+
+  return matchedNodes.sort((a, b) => a.nodeIndex - b.nodeIndex);
+}
+
+/**
+ * Add nodes to the passed in Set that match just a part of the selector in order to speed up traversing the entire tree.
+ * @param {Object} expression Selector Expression
+ * @param {Object} selectorMap Selector map cache
+ * @param {String} shadowId ShadowID of the root node
+ * @param {Set} nodeSet
+ */
+function findMatchingNodes(expression, selectorMap, shadowId, nodeSet) {
+  // use the last part of the expression to find nodes as it's more
+  // specific. e.g. for `body h1` use `h1` and not `body`
+  const exp = expression[expression.length - 1];
+  let nodes = [];
```

Comment:
Since the function no longer returns the empty array if passed in and only returns nodes that are in both arrays, it fixed the issue at the same time.

At step 3 when nothing is found, the empty array is passed to step 4, but since there are no nodes in the array the script won't add any nodes from the `.foo` array.

```
function getSharedValues(a, b) {
  return a.filter(node => b.includes(node));
}

getSharedValues([], ['a', 'b', 'c']);  // []
getSharedValues(['a', 'b', 'c'], []);  // []
```

Reply:
I don't think it is....? If somewhere along this algorithm you end up with the array getting emptied again, because of the ternary, you're not calling getSharedValues. You're doing `nodes = cachedNodes`, effectively.

This doesn't feel like a safe algorithm. It's hard to track. If you want this setup, have `nodes` be `null` until it is first assigned, and have your ternary be:

```js
 nodes = nodes === null ? cachedNodes : getSharedValues(nodes, cachedNodes);
```

Reply:
I see what you mean. Added a test to catch this problem now.